trieste/models/interfaces.py

```diff
@@ -241,6 +264,105 @@ def log(self) -> None:
             with tf.name_scope(f"{i}"):
                 model.log()
 
+    def reparam_sampler(self, num_samples: int) -> ReparametrizationSampler:
+        """
+        Return a reparametrization sampler providing `num_samples` samples across
+        all the models in the model stack.
+
+        :param num_samples: The desired number of samples.
+        :return: The reparametrization sampler.
+        :raise NotImplementedError: If the models in the stack do not share the
+            same :meth:`reparam_sampler`.
+        """
+
+        sampler_types = [type(model.reparam_sampler(1)) for model in self._models]
```

Comment:
As discussed, this optimization is only guaranteed to work for BatchReparameterizationSamplers, so you should restrict it to that. Also if possible you should definitely fall back to generating all the reparameterization samplers individually and  creating a stack sampler that combines the samples. 

Furthermore, since sampler initialisation is cheap for BatchReparameterizationSamplers but won't necessarily be for or other types, you might as well generate the actual samples when checking:

```suggestion
        samplers = [model.reparam_sampler(num_samples) for model in self._models]
        if all(isinstance(sampler, BatchReparameterizationSampler) for sampler in samplers):
            return BatchReparameterizationSampler(num_samples, self)
        else:
            ...
```

Reply:
Made your change. Will leave the general solution for stacks of models with different samplers for when we include another type of reparam sampler and can see what it actually looks like!